File: typescript/agent.ts
/**
 * Report service agent — TypeScript example.
 *
 * Listens for intents via SSE, generates reports, resumes with result.
 *
 * Usage:
 *   export AXME_API_KEY="<agent-key>"
 *   npx tsx agent.ts
 */

import { AxmeClient } from "@axme/axme";

const AGENT_ADDRESS = "report-service-demo";

async function handleIntent(client: AxmeClient, intentId: string) {
  const intentData = await client.getIntent(intentId);
  const intent = intentData.intent ?? intentData;
  let payload = intent.payload ?? {};
  if (payload.parent_payload) {
    payload = payload.parent_payload;
  }

  const reportType = payload.report_type ?? "unknown";
  const fmt = payload.format ?? "pdf";
  const quarter = payload.quarter ?? "Q1";

  console.log(`  Generating ${reportType} report (${fmt}) for ${quarter}...`);
  await new Promise((r) => setTimeout(r, 2000));

  const result = {
    action: "complete",
    report_url: `https://reports.example.com/${quarter}-${reportType}.${fmt}`,
    pages: 42,
    generated_at: new Date().toISOString(),
  };

  await client.resumeIntent(intentId, result, { ownerAgent: "report-service-demo" });
  console.log(`  Report ready: ${result.report_url}`);
}

async function main() {
  const apiKey = process.env.AXME_API_KEY;
  if (!apiKey) {
    console.error("Error: AXME_API_KEY not set.");
    process.exit(1);
  }

  const client = new AxmeClient({ apiKey });

  console.log(`Agent listening on ${AGENT_ADDRESS}...`);
  console.log("Waiting for intents (Ctrl+C to stop)\n");

  for await (const delivery of client.listen(AGENT_ADDRESS)) {
    const intentId = delivery.intent_id;
    const status = delivery.status;

    if (!intentId) continue;

    if (["DELIVERED", "CREATED", "IN_PROGRESS"].includes(status)) {
      console.log(`[${status}] Intent received: ${intentId}`);
      try {
        await handleIntent(client, intentId);
      } catch (e) {
        console.error(`  Error processing intent: ${e}`);
      }
    }
  }
}

main().catch(console.error);

File: agent.py
"""
Report service agent - listens for intents via SSE and processes them.

This agent simulates a report generation service. It receives an intent,
"generates" a report (simulated delay), and resumes with the result.

Usage:
    # Get agent key from scenario-agents.json after running scenario
    export AXME_API_KEY="<agent-key>"
    python agent.py
"""

import os
import sys
import time

# Ensure output is not buffered
sys.stdout.reconfigure(line_buffering=True)

from axme import AxmeClient, AxmeClientConfig


AGENT_ADDRESS = "report-service-demo"


def handle_intent(client, intent_id):
    """Process a single intent - generate report and resume."""
    intent_data = client.get_intent(intent_id)
    intent = intent_data.get("intent", intent_data)
    payload = intent.get("payload", {})
    # Scenario-based intents wrap original payload in parent_payload
    if "parent_payload" in payload:
        payload = payload["parent_payload"]

    report_type = payload.get("report_type", "unknown")
    fmt = payload.get("format", "pdf")
    quarter = payload.get("quarter", "Q1")

    print(f"  Generating {report_type} report ({fmt}) for {quarter}...")
    time.sleep(2)  # Simulate report generation

    result = {
        "action": "complete",
        "report_url": f"https://reports.example.com/{quarter}-{report_type}.{fmt}",
        "pages": 42,
        "generated_at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
    }

    client.resume_intent(intent_id, result)
    print(f"  Report ready: {result['report_url']}")


def main():
    api_key = os.environ.get("AXME_API_KEY", "")
    if not api_key:
        print("Error: AXME_API_KEY not set.")
        print("Run the scenario first: axme scenarios apply scenario.json")
        print("Then get the agent key from ~/.config/axme/scenario-agents.json")
        sys.exit(1)

    client = AxmeClient(AxmeClientConfig(api_key=api_key))

    print(f"Agent listening on {AGENT_ADDRESS}...")
    print("Waiting for intents (Ctrl+C to stop)\n")

    for delivery in client.listen(AGENT_ADDRESS):
        intent_id = delivery.get("intent_id", "")
        status = delivery.get("status", "")

        if not intent_id:
            continue

        # Process actionable deliveries
        if status in ("DELIVERED", "CREATED", "IN_PROGRESS"):
            print(f"[{status}] Intent received: {intent_id}")
            try:
                handle_intent(client, intent_id)
            except Exception as e:
                print(f"  Error processing intent: {e}")


if __name__ == "__main__":
    main()
